File: surf-chat/src/main/java/dev/slne/surf/social/chat/command/argument/ChannelMembersArgument.java
package dev.slne.surf.social.chat.command.argument;

import dev.jorel.commandapi.arguments.ArgumentSuggestions;
import dev.jorel.commandapi.arguments.CustomArgument;
import dev.jorel.commandapi.arguments.StringArgument;
import dev.slne.surf.social.chat.object.Channel;
import dev.slne.surf.social.chat.provider.ChannelProvider;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Bukkit;
import org.bukkit.OfflinePlayer;

public class ChannelMembersArgument extends CustomArgument<OfflinePlayer, String> {
  public ChannelMembersArgument(String nodeName) {
    super(new StringArgument(nodeName), info -> {
      OfflinePlayer player = Bukkit.getOfflinePlayer(info.input());
      Channel channel = Channel.getChannelO(player);

      if(channel == null) {
        throw CustomArgumentException.fromMessageBuilder(new MessageBuilder("Du bist in keinem Kanal, oder dieser ist invalid."));
      }

      if(!channel.isMember(player) && !channel.isOwnerO(player) && !channel.isModeratorO(player)) {
        throw CustomArgumentException.fromMessageBuilder(new MessageBuilder("Player is not a member of the channel: ").appendArgInput());
      }

      return player;
    });

    this.replaceSuggestions(ArgumentSuggestions.strings(info -> {
      try {
        return ChannelProvider.getInstance().getChannels().values()
            .stream()
            .filter(channel ->  channel.isMember(info.sender()) || channel.isOwner(info.sender())).findFirst().orElseThrow(() -> CustomArgumentException.fromString("You are not a member of any channel."))
            .getMembers()
            .stream()
            .filter(player -> player != info.sender())
            .map(OfflinePlayer::getName)
            .toArray(String[]::new);
      } catch (CustomArgumentException e) {
        info.sender().sendMessage(Component.text("You are not a member of any channel.").color(NamedTextColor.RED));
        return new String[0];
      }
    }));
  }
}

File: surf-chat/src/main/java/dev/slne/surf/social/chat/command/argument/ChannelMembersArgument.kt
package dev.slne.surf.social.chat.command.argument

import dev.jorel.commandapi.arguments.ArgumentSuggestions
import dev.jorel.commandapi.arguments.CustomArgument
import dev.jorel.commandapi.arguments.StringArgument
import dev.slne.surf.social.chat.`object`.Channel
import dev.slne.surf.social.chat.util.mapOfflinePlayerNamesTo
import dev.slne.surf.surfapi.core.api.util.emptyObjectSet
import dev.slne.surf.surfapi.core.api.util.mutableObjectSetOf
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.future.future
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer

@OptIn(DelicateCoroutinesApi::class)
class ChannelMembersArgument(nodeName: String) :
    CustomArgument<OfflinePlayer, String>(
        StringArgument(nodeName),
        { info ->
            val player = Bukkit.getOfflinePlayer(info.input())
            val uniqueId = player.uniqueId
            val channel = Channel.getChannel(uniqueId)
                ?: throw CustomArgumentException.fromMessageBuilder(MessageBuilder("Du bist in keinem Kanal, oder dieser ist invalid."))

            if (!channel.isMember(uniqueId) && !channel.isOwner(uniqueId)
                && !channel.isModerator(uniqueId)
            ) {
                throw CustomArgumentException.fromMessageBuilder(MessageBuilder("Player is not a member of the channel: ").appendArgInput())
            }

            player
        }) {
    init {
        this.replaceSuggestions(ArgumentSuggestions.stringCollectionAsync { info ->
            GlobalScope.future(Dispatchers.IO) {
                val channel =
                    Channel.getChannel(info.sender()) ?: return@future emptyObjectSet()
                val owner = channel.owner ?: return@future emptyObjectSet()

                mutableObjectSetOf<String>().apply {
                    channel.members.mapOfflinePlayerNamesTo(this)
                    channel.moderators.mapOfflinePlayerNamesTo(this)
                    Bukkit.getOfflinePlayer(owner).name?.let { add(it) }
                    remove(info.sender().name)
                }
            }
        })
    }
}